### ISSUE TITLE
Implementation of Trait for different type of Fn doesn't work

### ISSUE THREAD

**Issue by richerarc** [rust-lang/rust#60074]
Implementing a Trait for a type T (where T is a function with a given a signature) and then trying to implement the same trait for another type T (where T is a function with a different signature) doesn't compile.

I tried this code:
```Rust
trait MyTrait {}

impl<F> MyTrait for F where F: Fn(u32) -> u32 {}

impl<F> MyTrait for F where F: Fn(u32, u32) -> u32 {}
```

When compiling this code I get :
```
error[E0119]: conflicting implementations of trait `MyTrait`:
 --> src/main.rs:5:1
  |
3 | impl<F> MyTrait for F where F: Fn(u32) -> u32 {}
  | ------------------------------------------------------ first implementation here
4 | 
5 | impl<F> MyTrait for F where F: Fn(u32, u32) -> u32 {}
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation
```

What I would expect is that T here represent in the first implementation either a fn(u32) -> u32 or a closure the implement Fn(u32) -> u32, then in the second implementation, T would be either a fn(u32, u32) -> u32 or a closure the implement Fn(u32, u32) -> u32.

## Meta

rustc 1.34.0 (91856ed52 2019-04-10)
binary: rustc
commit-hash: 91856ed52c58aa5ba66a015354d1cc69e9779bdf
commit-date: 2019-04-10
host: x86_64-apple-darwin
release: 1.34.0
LLVM version: 8.0

**Comment by jonas-schievink** [comment]
I'm pretty sure this is a known issue, but can't find an issue to reference right now

**Comment by Nemo157** [comment]
This is an issue because a closure _could_ implement both `Fn` traits (even though they never will), e.g. using some nightly features you [can manually implement](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9cd8ba76e9d6d9dc3c58a1488f5ab8d0) the traits for a custom type:

```
impl FnOnce<(u32,)> for Foo {
    type Output = u32;
    extern "rust-call" fn call_once(self, args: (u32,)) -> Self::Output {
        args.0
    }
}

impl FnOnce<(u32, u32)> for Foo {
    type Output = u32;
    extern "rust-call" fn call_once(self, args: (u32, u32)) -> Self::Output {
        args.0
    }
}
```

There might be some way to make this work in the future with a new language feature, there's probably an issue about it open somewhere.

**Comment by Centril** [comment]
> There might be some way to make this work in the future with a new language feature, there's probably an issue about it open somewhere.

For example through variadic closures.

**Comment by Nemo157** [comment]
@Centril I was more thinking of something like forcing `Fn` traits to be mutually exclusive (maybe could even be accomplished by something as simple as moving `Args` to be an associated type), or some other way of propagating the fact that it's impossible to have an actual closure implement multiple versions of these traits (other than having generic implementations for HRTB etc.).

Alternatively something related to specialization so that both implementations can be provided and overlap could be dealt with.

But maybe variadic closures could do it as well.

**Comment by Centril** [comment]
@Nemo157 Sure; I was mostly thinking in terms of what can be done without a stable ability to implement these `Fn` traits directly... :)

**Comment by djugei** [comment]
Is there some workaround available for this issue? The best i can come up with right now is to wrap the functions in newtypes, which is slightly uncomfortable.

**Comment by aWeinzierl** [comment]
> I'm pretty sure this is a known issue, but can't find an issue to reference right now

https://github.com/rust-lang/rust/issues/20770

**Comment by dzmitry-lahoda** [comment]
@djugei was you able to wrap? i tried to wrap with into and from impls, but it said to me that somebody already impls these.

**Comment by ibraheemdev** [comment]
Why is this considered conflicting implementations?
```rust
trait Function {
    fn call(&self) -> i32;
}

impl<F: Fn() -> Result<i32, bool>> Function for F {
    fn call(&self) -> i32 {
        self().unwrap()
    }
}

impl<F: Fn() -> i32> Function for F {
    fn call(&self) -> i32 {
        self()
    }
}

// error[E0119]: conflicting implementations of trait `Function`
```

@Nemo157 mentioned that the original example is an is an issue because a closure could implement both `Fn` traits. However, in the above example, that would not be possible, because `Fn` is generic over `Args`, making it impossible to implement `Fn` twice with the same arguments:
```rust
struct Foo;

impl FnOnce<()> for Foo {
    type Output = i32;
    extern "rust-call" fn call_once(self, args: ()) -> Self::Output {
        1
    }
}

impl FnOnce<()> for Foo {
    type Output = Result<i32, bool>;
    extern "rust-call" fn call_once(self, args: ()) -> Self::Output {
        Ok(1)
    }
}

// error[E0119]: conflicting implementations of trait `std::ops::FnOnce<()>` for type `Foo`
```

**Comment by zserik** [comment]
@ibraheemdev yes, this is weird... https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f36063e2e6f332af7b6e261d3667bcfe

**Comment by geo-ant** [comment]
I saw that actix web has a solution for implementing a common interface across different `Fn` types. I was interested in doing a similiar thing for a project of mine, so I took their solution and [documented the crafty trick they used](https://geo-ant.github.io/blog/2021/rust-traits-and-variadic-functions/). If anyone's interested.
